packages/gatsby/src/schema/node-model.js

```diff
@@ -222,15 +222,29 @@ class LocalNodeModel {
   /**
    * Get nodes of a type matching the specified query.
    *
-   * @param {Object} args
-   * @param {Object} args.query Query arguments (`filter` and `sort`)
-   * @param {(string|GraphQLOutputType)} args.type Type
-   * @param {boolean} [args.firstOnly] If true, return only first match
-   * @param {PageDependencies} [pageDependencies]
-   * @returns {Promise<Node[]>}
+   * When `args.firstOnly` is true - behavior is exactly the same as `findOne`
+   *
+   * When `args.firstOnly` is falsy - behaves like `findAll` but returns an array
+   * instead of instance of queryResult, and ignores `args.query.limit` and `args.query.skip`
+   * (returns full result set, which is slow with LMDB).
+   *
+   * @deprecated Use `findAll` or `findOne` instead
    */
   async runQuery(args, pageDependencies = {}) {
-    const { query, firstOnly, type, stats, tracer } = args || {}
+    // TODO: show deprecation warning in v4
+    // reporter.warn(
+    //   `nodeModel.runQuery() is deprecated. Use nodeModel.findAll() or nodeModel.findOne() instead`
+    // )
+    if (args.firstOnly) {
+      return this.findOne(args, pageDependencies)
+    }
+    const { skip, limit, ...query } = args.query
+    const result = await this.findAll({ ...args, query }, pageDependencies)
+    return Array.from(result.entries)
```

Comment:
This is the main reason to deprecate `runQuery` - right now it ignores `limit`/`skip` and always returns full array of results.